experimental/torch_xla2/examples/train_llama_torchtitan/train_llama.py

```diff
@@ -232,37 +219,131 @@ def main(
     seqlen=2048,
-    mode='regular', 
+    override_num_layers=-1,
+    use_scan = True,
+    tp_parallelism=1,
 ):
-    logging.getLogger("jax").setLevel(logging.DEBUG)
+    torch_xla2.enable_globally()
+    torch_xla2.enable_performance_mode()
+    #logging.getLogger("jax").setLevel(logging.DEBUG)
     print(f"Running with parameters {locals()}")
-    global SEQLEN
-    global BATCH
-    SEQLEN = seqlen
-    BATCH = batch_size
 
-    mesh = jax.make_mesh((len(jax.local_devices()), ), ('fsdp', ))
+    fsdp = num_global_devices // tp_parallelism
+    mesh = jax.make_mesh((fsdp, tp_parallelism), ('fsdp', 'tp'))
+    if use_scan:
+        # using scan the individial weights will have shape (num_layers, w, h)
+        sharding_map = sharding_map_scan_fsdp
+    else:
+        sharding_map = sharding_map_original
+
     env = torch_xla2.default_env()
-    env.config.use_tpu_flash_attention = use_flash_attention
-    env.config.shmap_flash_attention = use_flash_attention
+    env.config.use_tpu_flash_attention = True
+    env.config.shmap_flash_attention = True
+    env._mesh = mesh  # this is the mesh used by flash attention pallas kernel
 
     args = llama3_configs[model_type]
-    #with torch.device('meta'):
-    gpt = titan.Transformer(args)
-
-    light_mod = Module(gpt)
-    light_mod.to(torch.bfloat16)
-
+    # Note: torchtitan's upstream config did not specify this value
+    args.vocab_size = 128256
+    args.max_seq_len = seqlen
+    if override_num_layers > 0:
+        args.n_layers = override_num_layers
+
+    # Note: because a single device don't have enough HBM memory
+    # nor enough CPU memory to hold the parameters. We instantiate
+    # the model on meta then manually initialize then shard each param
+    torch.set_default_dtype(torch.bfloat16)
+    with torch.device('meta'):
+        gpt = titan.Transformer(args)
+        
+    with torch.device('cpu'):
+        # need actual value for freqs_cis
+        freqs_cis = gpt._precompute_freqs_cis()
+
+    if use_scan:
+        checkpoint_policy=jax.checkpoint_policies.offload_dot_with_no_batch_dims('device', 'pinned_host')
+        gpt = TransfomerWithScan(gpt, checkpoint_policy) 
+
+    state_dict = dict(gpt.state_dict())
+    state_dict.pop('freqs_cis') # dont shard freqs_cis
+    state_dict = create_sharded_weights(gpt, mesh, sharding_map)
+    replicated = jax.sharding.NamedSharding(mesh, P())
+
+    state_dict['freqs_cis'] = freqs_cis.to('jax').apply_jax(jax.device_put, replicated)
+    gpt.load_state_dict(state_dict, assign=True)
+    
     train_loader = fake_dataloader(10, seqlen, batch_size)
 
+    
+    # NOTE: overriding attention to capture mesh and sharding info
+    partition = P('fsdp', 'tp', None, None)
+    attention = functools.partial(
+      splash_attn.tpu_splash_attention, 
+      mesh, partition, True)
+    attention = jax.jit(attention)
+
+    def custom_attention(
+        query, key, value, attn_mask=None,
+        dropout_p=0.0, is_causal=False,
```

Comment:
What happens if `dropout_p` is not zero or when `is_causal` is `False`? Should we assert that their values matches the behavior of `splash_attention`?

Reply:
This bit is to show case that you can register your own override of op.
So the user is writing this bit and we assume the user would know which special case it applies to them.